### ISSUE TITLE
Disable multi-threading?

### ISSUE THREAD

**Issue by legolord208** [gyscos/Cursive#140]
Hello! I am a Rust beginner, so excuse me if I say something really unintelligent.

I am having difficulties understanding why Multi-threading is needed, and also I am having some very big difficulties understanding how to access outside variables. See https://www.reddit.com/r/rust/comments/6bv3as/help_variables_closures_headache/?ref=share&ref_source=link.

I'd appreciate if it would be possible to disable it? Cheers!

**Comment by gyscos** [comment]
Hi, and thanks for the report!

Are you still having a problem? (You mention it's solved on reddit.)
In general, cursive doesn't force any multi-threading, and I didn't see threads in the linked post. If this is still a problem, could you describe the context for the function you posted?

**Comment by legolord208** [comment]
I just assumed it was multi-threaded since the kind of `outlive the current function` errors is caused by threads?

Also, it says on the README :stuck_out_tongue:

--------------------------------------

But yeah, I solved the problem itself. Still, it would be nice to do it without using mutexes and stuff.

**Comment by gyscos** [comment]
The UI mentions multi-threaded UI as an explicit non-goal... :-p

The `outlive the current function` error message is specific to closures, not just threads. Indeed, the closure sometimes ends up "living" longer than the function it is created in (when you give it as a callback).

You're using `Arc<Mutex<...>>`, which is the thread-safe version of `Rc<RefCell<...>>`.
You need this because the compiler cannot *prove* that the context is not in use by someone else while you edit it. For instance, in a different program, one button could borrow the context, then trigger another button click, which could then borrow the context again.
Since it's hard to detect this at compile-time, we have to use runtime checks instead to enforce that the context is only used in one place at a time. This is what `RefCell` does.

TLDR: you don't *need* `Mutex`, but you need *something*. Replacing `Arc<Mutex<...>>` with `Rc<RefCell<...>>` will at least improve performance and get rid of the multi-thread-specific code. It's a common pattern when something (here, the context) has to be accessed from multiple places.

**Comment by legolord208** [comment]
Holy shit you are the kindest person on Earth! I'll close the issue since it's not related to your library - sorry for disturbing.
